### PR TITLE
Add apache-arrow-glib formula

### PR DESCRIPTION
**Pull request by mrkn**
This provides GObject Introspection files of Apache Arrow

### CODE REVIEW

**Comment by cosmo0920** [comment]
I think that this formula should be bundled into apache-arrow formula.

**Comment by mrkn** [comment]
I've already tried to integrate C API into apache-arrow formula, but I couldn't due to rpath related problem.  I got the following build error at the task to generate Arrow-1.0.gir file:

```
dyld: Library not loaded: @rpath/libarrow.0.dylib
  Referenced from: /private/tmp/apache-arrow-20170927-27674-j5qc9z/apache-arrow-0.7.0/c_glib/arrow-glib/tmp-introspectuDHZB4/.libs/Arrow-1.0
  Reason: image not found
Command '[u'/private/tmp/apache-arrow-20170927-27674-j5qc9z/apache-arrow-0.7.0/c_glib/arrow-glib/tmp-introspectuDHZB4/Arrow-1.0', u'--introspect-dump=/private/tmp/apache-arrow-20170927-27674-j5qc9z/apache-arrow-0.7.0/c_glib/arrow-glib/tmp-introspectuDHZB4/functions.txt,/private/tmp/apache-arrow-20170927-27674-j5qc9z/apache-arrow-0.7.0/c_glib/arrow-glib/tmp-introspectuDHZB4/dump.xml']' returned non-zero exit status -6
make[3]: *** [Arrow-1.0.gir] Error 1
```

Moreover, separating apache-arrow-glib doesn't force to let the original apache-arrow package depend on glib and gobject-introspection who are not required just to use Apache Arrow.

File: Formula/apache-arrow-glib.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      system "make"
+      system "make", "install"
+    end
+  end
+
+  test do
+    ENV['GEM_HOME'] = testpath/'gems'
+    system "gem", "install", "gobject-introspection"
+    system "ruby", "-rgi", "-e", "GI.load('Arrow')"
+  end
```
Comment (cosmo0920) [inline Formula/apache-arrow-glib.rb]: How about adding C code test?

```ruby
  test do
    (testpath/"test.c").write <<-EOS.undent
      #include "arrow-glib/arrow-glib.h"
      int main(void)
      {
        GArrowInt32ArrayBuilder *builder;
        return 0;
      }
    EOS
    glib = Formula["glib"]
    flags = (ENV.cflags || "").split + (ENV.cppflags || "").split + (ENV.ldflags || "").split
    flags += %W[
      -I#{include}
      -I#{glib.opt_include}/glib-2.0
      -I#{glib.opt_lib}/glib-2.0/include
      -L#{lib}
      -lglib-2.0
      -larrow-glib
    ]
    system ENV.cc, "test.c", "-o", "test", *flags
    system "./test"
  end
```

Comment (mrkn) [reply Formula/apache-arrow-glib.rb]: Can't we examine gobject-introspection things with C code test?

Comment (cosmo0920) [reply Formula/apache-arrow-glib.rb]: For functionality test, we should use gi-based apache arrow glib binding in testing.
But for homebrew formula test, we should verify just installation about headers and libraries.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Review by tschoonj — APPROVED** [review]
Looks good to me. The test is fine too by the way since the idea with gobject-introspection is to use C libraries from within other languages.

**Comment by mrkn** [comment]
The issues on CI have been resolved, so this pull-request is ready to merge.

**Comment by mrkn** [comment]
Is there anything I should do before merging this?

**Comment by mrkn** [comment]
I updated the commit against apache-arrow 0.7.1.

File: Formula/apache-arrow-glib.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+class ApacheArrowGlib < Formula
+  desc "GObject Introspection files of Apache Arrow"
+  homepage "https://arrow.apache.org/"
+  url "https://www.apache.org/dyn/closer.cgi?path=arrow/arrow-0.7.1/apache-arrow-0.7.1.tar.gz"
+  sha256 "f8f114d427a8702791c18a26bdcc9df2a274b8388e08d2d8c73dd09dc08e888e"
+  head "https://github.com/apache/arrow.git"
+
+  depends_on "apache-arrow" => "0.7.1"
```
Comment (ilovezfs) [inline Formula/apache-arrow-glib.rb]: We don't support depending on specific versions

**Review by ilovezfs — DISMISSED** [review]

File: Formula/apache-arrow-glib.rb (hunk 2 above)
Comment (mrkn) [reply Formula/apache-arrow-glib.rb]: So, is there no way to force the version of depending apache-arrow?

Comment (ilovezfs) [reply Formula/apache-arrow-glib.rb]: Nope. It has to use whatever the current version in the other formula is.

Comment (mrkn) [reply Formula/apache-arrow-glib.rb]: I fixed it.

**Comment by ilovezfs** [comment]
go away bot

File: Formula/apache-arrow-glib.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      system "./configure", *args
+      system "make"
+      system "make", "install"
+    end
+  end
+
+  test do
+    ENV["ARCHFLAGS"] = ""
+    ENV["GEM_HOME"] = testpath/"gems"
+    gem = "gem"
```
Comment (ilovezfs) [inline Formula/apache-arrow-glib.rb]: You can just use `system "ruby"` and `system "gem"` now.

File: Formula/apache-arrow-glib.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        CC=#{ENV.cc}
+      ]
+      system "./configure", *args
+      system "make"
+      system "make", "install"
+    end
+  end
+
+  test do
+    ENV["ARCHFLAGS"] = ""
```
Comment (ilovezfs) [inline Formula/apache-arrow-glib.rb]: Why?

File: Formula/apache-arrow-glib.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  depends_on "glib"
+  depends_on "gobject-introspection"
+
+  def install
+    cd "c_glib" do
+      args = %W[
+        --prefix=#{prefix}
+        CC=#{ENV.cc}
+      ]
+      system "./configure", *args
```
Comment (ilovezfs) [inline Formula/apache-arrow-glib.rb]: just inline the argument

```
      system "./configure", "--prefix=#{prefix}"
```

The `CC=#{ENV.cc}` shouldn't be needed.

File: Formula/apache-arrow-glib.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
+class ApacheArrowGlib < Formula
+  desc "GObject Introspection files of Apache Arrow"
+  homepage "https://arrow.apache.org/"
+  url "https://www.apache.org/dyn/closer.cgi?path=arrow/arrow-0.7.1/apache-arrow-0.7.1.tar.gz"
+  sha256 "f8f114d427a8702791c18a26bdcc9df2a274b8388e08d2d8c73dd09dc08e888e"
+  head "https://github.com/apache/arrow.git"
+
+  depends_on "apache-arrow"
+  depends_on "pkg-config" => :build
```
Comment (ilovezfs) [inline Formula/apache-arrow-glib.rb]: `=> :build` deps go above all others

File: Formula/apache-arrow-glib.rb (hunk 4 above)
Comment (mrkn) [reply Formula/apache-arrow-glib.rb]: I heard from @kou that `-arch i386` is supplied by Python's distutils without setting empty`ARCHFLAGS`, and then the build was failed.
So I specify the empty `ARCHFLAGS` here.

Comment (mrkn) [reply Formula/apache-arrow-glib.rb]: I don't have High Sierra environment, so I've never confirmed that by myself.
It is just in case.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

File: Formula/apache-arrow-glib.rb (hunk 4 above)
Comment (ilovezfs) [reply Formula/apache-arrow-glib.rb]: We filter that out automatically.

**Comment by ilovezfs** [comment]
@mrkn what's the status here?

**Comment by ilovezfs** [comment]
Feel free to re-open when the issues noted above have been addressed. Thanks @mrkn

**Pull request closed without merging**